libs/server/Objects/ItemBroker/CollectionItemBroker.cs

```diff
@@ -0,0 +1,578 @@
+                        switch (command)
+                        {
+                            case RespCommand.BLPOP:
+                            case RespCommand.BRPOP:
+                                return TryGetNextListItem(listObj, command, out nextItem);
+                            case RespCommand.BLMOVE:
+                                ListObject dstList;
+                                var newObj = false;
+                                if (dstObj == null)
+                                {
+                                    dstList = new ListObject();
+                                    newObj = true;
+                                }
+                                else if (dstObj is ListObject tmpDstList)
+                                {
+                                    dstList = tmpDstList;
+                                }
+                                else return false;
+
+                                var isSuccessful = TryMoveNextListItem(listObj, dstList, (OperationDirection)cmdArgs[1],
+                                    (OperationDirection)cmdArgs[2], out nextItem);
+
+                                if (isSuccessful && newObj)
+                                {
+                                    isSuccessful = storageSession.SET(dstKey, dstList, ref objectLockableContext) ==
+                                                   GarnetStatus.OK;
+                                }
+
+                                return isSuccessful;
+                            default:
+                                return false;
+                        }
+                    case SortedSetObject setObj:
+                        currCount = setObj.Dictionary.Count;
+                        if (objectType != GarnetObjectType.SortedSet) return false;
+                        return TryGetNextSetObject(setObj, command, out nextItem);
+                    default:
+                        return false;
+                }
+            }
+            finally
+            {
+                if (createTransaction)
+                    storageSession.txnManager.Commit(true);
+            }
+        }
+
+        /// <summary>
+        /// Broker's main loop logic
+        /// </summary>
+        /// <returns>Task</returns>
+        private async Task Start()
+        {
+            Task handleNextEvent = default;
+            try
+            {
+                // Repeat while not disposed or cancelled
+                while (!disposed && !cts.IsCancellationRequested)
+                {
+                    // Check if current task is done
+                    if (handleNextEvent == null || handleNextEvent.IsCompleted)
+                    {
+                        // Set task to asynchronously dequeue next event in broker's queue
+                        // once event is dequeued successfully, call handler method
+                        handleNextEvent = brokerEventsQueue.DequeueAsync(cts.Token).ContinueWith(t =>
```

Comment:
will there be a task allocation for every item in the queue? for heavy load, can we synchronously dequeue, and only when that fails we actually incur the cost of async?